cmd/ftl/cmd_open.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	errors "github.com/alecthomas/errors"
+
+	"github.com/block/ftl/backend/protos/xyz/block/ftl/admin/v1/adminpbconnect"
+	"github.com/block/ftl/backend/protos/xyz/block/ftl/buildengine/v1/buildenginepbconnect"
+	"github.com/block/ftl/common/reflection"
+	"github.com/block/ftl/common/schema"
+	"github.com/block/ftl/internal/devstate"
+	"github.com/block/ftl/internal/exec"
+	"github.com/block/ftl/internal/log"
+	"github.com/block/ftl/internal/projectconfig"
+)
+
+const (
+	autodetect = "auto"
+	vscode     = "vscode"
+	intellij   = "intellij"
+)
+
+type openCmd struct {
+	Ref    reflection.Ref `arg:"" help:"Language of the module to create." placeholder:"MODULE.ITEM" predictor:"decls"`
+	Editor string         `arg:"" help:"Editor to open the file with." enum:"auto,vscode,intellij" default:"auto"`
+
+	TerminalProgram  string `help:"Helps detect if the terminal is running within an IDE, used for 'auto' IDE detection" env:"TERM_PROGRAM" hidden:""`
+	TerminalEmulator string `help:"Helps detect if the terminal is running within an IDE, used for 'auto' IDE detection" env:"TERMINAL_EMULATOR" hidden:""`
+}
+
+func (i openCmd) Run(ctx context.Context, buildEngineClient buildenginepbconnect.BuildEngineServiceClient, adminClient adminpbconnect.AdminServiceClient, pc projectconfig.Config) error {
+	// Currently dev state is the easiest way to get schema and module paths
+	// We should use admin client to get schema directly when we have a single directory for modules in a project
+	state, err := devstate.WaitForDevState(ctx, buildEngineClient, adminClient, false)
+	if err != nil {
+		return errors.Wrapf(err, "could not get dev state")
+	}
+
+	odecl, _ := state.Schema.ResolveWithModule(&schema.Ref{Module: i.Ref.Module, Name: i.Ref.Name})
+	decl, ok := odecl.Get()
+	if !ok {
+		return errors.Errorf("could not find %q", i.Ref)
+	}
+	if decl.Position().Filename == "" {
+		return errors.Errorf("could not find file of %q", i.Ref)
+	}
+
+	if i.Editor == autodetect {
+		if i.TerminalProgram == "vscode" {
+			i.Editor = vscode
+		} else if strings.Contains(i.TerminalEmulator, "JetBrains") {
+			i.Editor = intellij
+		} else {
+			return errors.Errorf("could not auto choose default editor, expected one of %s", strings.Join([]string{vscode, intellij}, ", "))
+		}
+	}
```

Comment:
Yeah can we pull this out into a dedicated package and make it extensible instead?

Reply:
Or maybe just move this into the user's config?

It's very difficult and non-cross-platform to open an editor automatically, so maybe we just punt on it...

For example, this won't work for me at all as I don't use vscode or JetBrains.

Reply:
Yeh puntings easiest